src/bin/button.rs

```diff
@@ -0,0 +1,268 @@
+        loop {
+            match embassy_time::with_timeout(
+                embassy_time::Duration::from_secs(15),
+                self.0.changed(),
+            )
+            .await
+            {
+                Ok(value) => {
+                    writer
+                        .write_event("value_changed", value.to_string().as_str())
+                        .await?
+                }
+                Err(_) => writer.write_keepalive().await?,
+            }
+        }
+    }
+}
+
+#[esp_hal_embassy::main]
+async fn main(spawner: Spawner) {
+    esp_println::logger::init_logger_from_env();
+    let peripherals = esp_hal::init({
+        let mut config = esp_hal::Config::default();
+        config.cpu_clock = CpuClock::max();
+        config
+    });
+
+    esp_alloc::heap_allocator!(72 * 1024);
+
+    let timg0 = TimerGroup::new(peripherals.TIMG0);
+
+    let init = &*mk_static!(
+        EspWifiController<'static>,
+        init(
+            timg0.timer0,
+            Rng::new(peripherals.RNG),
+            peripherals.RADIO_CLK,
+        )
+        .unwrap()
+    );
+
+    let wifi = peripherals.WIFI;
+    let (wifi_interface, controller) =
+        esp_wifi::wifi::new_with_mode(&init, wifi, WifiStaDevice).unwrap();
+
+    use esp_hal::timer::systimer::{SystemTimer, Target};
+    let systimer = SystemTimer::new(peripherals.SYSTIMER).split::<Target>();
+    esp_hal_embassy::init(systimer.alarm0);
+
+    let config = embassy_net::Config::dhcpv4(Default::default());
+
+    let seed = 1234; // very random, very secure seed
+
+    // Init network stack
+    let stack = &*mk_static!(
+        Stack<WifiDevice<'_, WifiStaDevice>>,
+        Stack::new(
+            wifi_interface,
+            config,
+            mk_static!(StackResources<3>, StackResources::<3>::new()),
+            seed
+        )
+    );
+
+    spawner.spawn(connection(controller)).ok();
+    spawner.spawn(net_task(&stack)).ok();
+
+    loop {
+        if stack.is_link_up() {
+            break;
+        }
+        Timer::after(Duration::from_millis(500)).await;
+    }
+
+    let base_uri;
+    println!("Waiting to get IP address...");
+    loop {
+        if let Some(config) = stack.config_v4() {
+            println!("Got IP: {}", config.address);
+            base_uri = format!("http://{}", config.address.address());
+            break;
+        }
+        Timer::after(Duration::from_millis(500)).await;
+    }
+
+    let btn = Input::new(peripherals.GPIO9, Pull::Up);
+
+    let id = get_urn_or_uuid(stack);
+
+    let td = Thing::builder("shtc3")
+        .finish_extend()
+        .id(id)
+        .base(base_uri)
+        .description("Example Thing exposing a shtc3 sensor")
+        .security(|builder| builder.no_sec().required().with_key("nosec_sc"))
+        .property("on", |p| {
+            p.finish_extend_data_schema()
+                .attype("TemperatureProperty")
+                .title("Temperature")
+                .description("Current temperature")
```

Comment:
I think these probably need to adjusted, right?

Reply:
Indeed :)